File: src/h1tdfsdiameter/JS/Main.js
import { TreeNode } from './TreeNode.js';
import { BinaryTree } from './BinaryTree.js';

// Function to display the tree structure (simple level-order traversal for illustration)
function displayTree(root) {
    if (root === null) {
        return;
    }
    let queue = [root];
    while (queue.length > 0) {
        let curr = queue.shift();
        process.stdout.write(curr.data + " ");
        if (curr.left !== null) {
            queue.push(curr.left);
        }
        if (curr.right !== null) {
            queue.push(curr.right);
        }
    }
    console.log();
}

function main() {
    const listOfTrees = [
        [new TreeNode(2), new TreeNode(1), new TreeNode(4), new TreeNode(3), new TreeNode(5), new TreeNode(6), new TreeNode(7)],
        [new TreeNode(1), new TreeNode(2), new TreeNode(3), new TreeNode(4), new TreeNode(5), new TreeNode(6), new TreeNode(7), new TreeNode(8), new TreeNode(9)],
        [new TreeNode(45), new TreeNode(32), new TreeNode(23), new TreeNode(21), new TreeNode(19), new TreeNode(19), new TreeNode(18), new TreeNode(1)],
        [new TreeNode(5), new TreeNode(3), new TreeNode(4), new TreeNode(1), new TreeNode(2), new TreeNode(6), new TreeNode(7), new TreeNode(8), new TreeNode(9)],
        [new TreeNode(9), new TreeNode(7), null, null, new TreeNode(1), new TreeNode(8), new TreeNode(10), null, new TreeNode(12)]
    ];

    const inputTrees = listOfTrees.map(listOfNodes => new BinaryTree(listOfNodes));

    inputTrees.forEach((tree, index) => {
        console.log(`${index + 1}.\tOriginal tree:`);
        displayTree(tree.root);
        console.log(`\tDiameter of tree: ${tree.diameterOfBinaryTree()}`);
        console.log('-'.repeat(100));
    });
}

main();

File: src/h1tdfsdiameter/Go/Main.go
package main

import (
	"fmt"
	"strings"
)

// displayTree prints the tree structure (a simple example for illustration)
func displayTree[T any](root *TreeNode[T]) {
	if root == nil {
		return
	}
	queue := []*TreeNode[T]{root}
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		fmt.Print(curr.Data, " ")
		if curr.Left != nil {
			queue = append(queue, curr.Left)
		}
		if curr.Right != nil {
			queue = append(queue, curr.Right)
		}
	}
	fmt.Println()
}

func main() {
	listOfTrees := [][]*TreeNode[int]{
		{&TreeNode[int]{Data: 2}, &TreeNode[int]{Data: 1}, &TreeNode[int]{Data: 4}, &TreeNode[int]{Data: 3}, &TreeNode[int]{Data: 5}, &TreeNode[int]{Data: 6}, &TreeNode[int]{Data: 7}},
		{&TreeNode[int]{Data: 1}, &TreeNode[int]{Data: 2}, &TreeNode[int]{Data: 3}, &TreeNode[int]{Data: 4}, &TreeNode[int]{Data: 5}, &TreeNode[int]{Data: 6}, &TreeNode[int]{Data: 7}, &TreeNode[int]{Data: 8}, &TreeNode[int]{Data: 9}},
		{&TreeNode[int]{Data: 45}, &TreeNode[int]{Data: 32}, &TreeNode[int]{Data: 23}, &TreeNode[int]{Data: 21}, &TreeNode[int]{Data: 19}, &TreeNode[int]{Data: 19}, &TreeNode[int]{Data: 18}, &TreeNode[int]{Data: 1}},
		{&TreeNode[int]{Data: 5}, &TreeNode[int]{Data: 3}, &TreeNode[int]{Data: 4}, &TreeNode[int]{Data: 1}, &TreeNode[int]{Data: 2}, &TreeNode[int]{Data: 6}, &TreeNode[int]{Data: 7}, &TreeNode[int]{Data: 8}, &TreeNode[int]{Data: 9}},
		{&TreeNode[int]{Data: 9}, &TreeNode[int]{Data: 7}, nil, nil, &TreeNode[int]{Data: 1}, &TreeNode[int]{Data: 8}, &TreeNode[int]{Data: 10}, nil, &TreeNode[int]{Data: 12}},
	}

	inputTrees := make([]*BinaryTree[int], len(listOfTrees))
	for i, listOfNodes := range listOfTrees {
		inputTrees[i] = NewBinaryTree(listOfNodes)
	}

	for i, tree := range inputTrees {
		fmt.Printf("%d.\tOriginal tree:\n", i+1)
		displayTree(tree.root)
		fmt.Printf("\n\tDiameter of tree: %d\n", tree.DiameterOfBinaryTree())
		fmt.Printf("%s\n", strings.Repeat("-", 100))
	}
}
